crates/ibc/src/core/ics02_client/client_state.rs

```diff
@@ -67,19 +68,54 @@ pub trait ClientState:
 
     fn initialise(&self, consensus_state: Any) -> Result<Box<dyn ConsensusState>, ClientError>;
 
-    fn check_header_and_update_state(
+    /// verify_client_message must verify a client_message. A client_message
+    /// could be a Header, Misbehaviour. It must handle each type of
+    /// client_message appropriately. Calls to check_for_misbehaviour,
+    /// update_state, and update_state_on_misbehaviour will assume that the
+    /// content of the client_message has been verified and can be trusted. An
+    /// error should be returned if the client_message fails to verify.
+    fn verify_client_message(
         &self,
         ctx: &dyn ValidationContext,
-        client_id: ClientId,
-        header: Any,
-    ) -> Result<UpdatedState, ClientError>;
+        client_id: &ClientId,
+        client_message: Any,
+        update_kind: &UpdateClientKind,
+    ) -> Result<(), ClientError>;
 
-    fn check_misbehaviour_and_update_state(
+    /// Checks for evidence of a misbehaviour in Header or Misbehaviour type. It
+    /// assumes the client_message has already been verified.
+    fn check_for_misbehaviour(
         &self,
         ctx: &dyn ValidationContext,
-        client_id: ClientId,
-        misbehaviour: Any,
-    ) -> Result<Box<dyn ClientState>, ContextError>;
+        client_id: &ClientId,
+        client_message: Any,
+        update_kind: &UpdateClientKind,
+    ) -> Result<bool, ClientError>;
+
+    /// Updates and stores as necessary any associated information for an IBC
+    /// client, such as the ClientState and corresponding ConsensusState. Upon
+    /// successful update, a list of consensus heights is returned. It assumes
+    /// the client_message has already been verified.
+    ///
+    /// Post-condition: on success, the return value MUST contain at least one
+    /// height.
+    fn update_state(
+        &self,
```

Comment:
consider making this `self` instead

Reply:
This doesn't work since `Self` is a `dyn ClientState`, and so cannot be moved (it must always live behind a reference).